src/java.base/share/classes/java/security/SecureRandom.java

```diff
@@ -267,23 +275,38 @@ public SecureRandom(byte[] seed) {
     private void getDefaultPRNG(boolean setSeed, byte[] seed) {
         Service prngService = null;
         String prngAlgorithm = null;
-        for (Provider p : Providers.getProviderList().providers()) {
-            // SUN provider uses the SunEntries.DEF_SECURE_RANDOM_ALGO
-            // as the default SecureRandom algorithm; for other providers,
-            // Provider.getDefaultSecureRandom() will use the 1st
-            // registered SecureRandom algorithm
-            if (p.getName().equals("SUN")) {
-                prngAlgorithm = SunEntries.DEF_SECURE_RANDOM_ALGO;
-                prngService = p.getService("SecureRandom", prngAlgorithm);
-                break;
-            } else {
-                prngService = p.getDefaultSecureRandomService();
-                if (prngService != null) {
-                    prngAlgorithm = prngService.getAlgorithm();
+
+        // If in FIPS mode, use the SecureRandom from the FIPS provider.
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prngAlgorithm = "PKCS11";
+            if (p == null) {
+                throw new RuntimeException("could not find a implementation for PKCS11 in SunPKCS11-NSS-FIPS.");
+            }
+            prngService = p.getService("SecureRandom", prngAlgorithm);
+            if (prngService == null) {
+                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider in FIPS mode.");
+            }
```

Comment:
Please don't include the period in the messages.

I think you have the messages swapped: the first should be about the missing provider, the second about the missing service.

Also correct the grammar (a -> an), or use my shorter suggestions:
```
    "could not find SunPKCS11-NSS-FIPS provider for FIPS mode"
```
and
```
   "could not find SecureRandom implementation from SunPKCS11-NSS-FIPS"
```

Reply:
updated